File: 20-actors/niyaku/methods/crane_dynamics.py
"""crane_dynamics — gantry / ship-to-shore (STS) crane anti-sway physics core.

The defining control problem of automated container handling is **anti-sway**:
a quay crane moves a 20-40 t container suspended on cables while the trolley
traverses 30-50 m ship→shore. The suspended load is a pendulum; an aggressive
trolley move excites residual sway that must settle to < a few cm at the target
before the spreader can land the box on a stack tier. This is the classical
*cart + hanging payload* model — the same topology as Isaac Sim's Cartpole
(prismatic trolley + revolute load), which is why ``isaac_sway_sim`` can drive
the load through the clean-room ``isaacsim.core.api`` surface.

This module is the **analytic/control core**: physically-correct hanging
pendulum, a state-feedback anti-sway position controller, and a ZV
input-shaper (the technique most container terminals actually deploy).

stdlib-only · pywasm-ready · no NumPy. All state is plain ``list``/``float``.

Per ADR-2606074000 (niyaku R0). Methods are pure compute (no outward action,
G-no-server-key); they do not move a real crane.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import List, Tuple

# Sign convention
#   x        : trolley position along the quay rail (m), shore-positive
#   theta    : load swing angle from vertical (rad); theta>0 ⇒ load lags +x
#   cable    : hoist cable length from trolley to load CG (m)
# Equilibrium is the load hanging straight down (theta = 0) — the STABLE point,
# unlike the inverted Cartpole. Gravity restores the load toward theta = 0.


@dataclass
class GantryCrane:
    """Reduced-order single-pendulum-on-trolley model of an STS / RTG crane.

    The trolley is treated as acceleration-commanded (its own servo loop is far
    faster than the sway mode), so the control input ``u`` is trolley
    acceleration (m/s²), saturated at ``accel_max``.
    """

    cable_length: float = 30.0      # m, spreader+load below trolley
    gravity: float = 9.81           # m/s²
    sway_damping: float = 0.02      # dimensionless viscous damping ratio proxy
    accel_max: float = 0.6          # m/s², trolley accel envelope (comfort/safety)
    velocity_max: float = 4.0       # m/s, trolley max traverse speed
    rail_length: float = 60.0       # m, usable trolley travel

    def natural_frequency(self) -> float:
        """Undamped sway natural frequency ω = sqrt(g / L) (rad/s)."""
        return math.sqrt(self.gravity / self.cable_length)

    def sway_period(self) -> float:
        """Sway period T = 2π / ω (s) — sets the input-shaper impulse spacing."""
        return 2.0 * math.pi / self.natural_frequency()

    # ── dynamics ────────────────────────────────────────────────────────────

    def derivatives(self, state: List[float], u: float) -> List[float]:
        """Continuous-time state derivative for state = [x, x_dot, theta, theta_dot].

        Full (non-linearised) hanging-pendulum-on-trolley with viscous sway
        damping. Trolley acceleration equals the (clamped) command ``u``.
        """
        x, x_dot, theta, theta_dot = state
        a = self._clamp(u, self.accel_max)
        L = self.cable_length
        g = self.gravity
        # theta'' = -(g/L) sinθ - (a/L) cosθ - 2ζω θ'
        zeta_w = self.sway_damping * self.natural_frequency()
        theta_acc = (
            -(g / L) * math.sin(theta)
            - (a / L) * math.cos(theta)
            - 2.0 * zeta_w * theta_dot
        )
        return [x_dot, a, theta_dot, theta_acc]

    def step(self, state: List[float], u: float, dt: float) -> List[float]:
        """Advance one step by classic RK4 (stable for the stiff sway mode)."""
        def add(s: List[float], k: List[float], h: float) -> List[float]:
            return [s[i] + h * k[i] for i in range(4)]

        k1 = self.derivatives(state, u)
        k2 = self.derivatives(add(state, k1, dt / 2.0), u)
        k3 = self.derivatives(add(state, k2, dt / 2.0), u)
        k4 = self.derivatives(add(state, k3, dt), u)
        nxt = [
            state[i] + (dt / 6.0) * (k1[i] + 2.0 * k2[i] + 2.0 * k3[i] + k4[i])
            for i in range(4)
        ]
        # enforce the trolley velocity envelope (servo limit)
        nxt[1] = self._clamp(nxt[1], self.velocity_max)
        return nxt

    @staticmethod
    def _clamp(v: float, lim: float) -> float:
        return max(-lim, min(lim, v))


# ── anti-sway state-feedback controller ─────────────────────────────────────


@dataclass
class AntiSwayController:
    """PD trolley positioning + sway-rate feedback.

    u = -kp (x - x_target) - kd x_dot + k_theta θ + k_thetad θ_dot

    The sway terms actively bleed pendulum energy so the load arrives quiet.
    Their sign is POSITIVE because here the equilibrium is the load hanging at
    θ=0 and trolley acceleration couples as -a/L into θ̈ (a forward push drives
    θ negative); a positive θ feedback therefore stiffens the restoring term
    (θ̈ ← -(g+k_theta)/L · θ) rather than cancelling it. Gains are normalised to
    the crane's natural frequency so a longer cable (slower sway) needs no
    re-tuning.
    """

    kp: float = 0.4
    kd: float = 1.7
    k_theta: float = 5.0
    k_thetad: float = 3.0

    def command(self, crane: GantryCrane, state: List[float], x_target: float) -> float:
        x, x_dot, theta, theta_dot = state
        w = crane.natural_frequency()
        u = (
            -self.kp * w * w * (x - x_target)
            - self.kd * w * x_dot
            + self.k_theta * theta
            + self.k_thetad / w * theta_dot
        )
        return crane._clamp(u, crane.accel_max)


# ── ZV input shaper (open-loop anti-sway, the terminal-deployed technique) ───


def zv_shaper(crane: GantryCrane) -> List[Tuple[float, float]]:
    """Zero-Vibration (ZV) input shaper impulses [(time_s, amplitude), ...].

    Two impulses spaced half a damped sway period cancel residual oscillation
    of a command applied through them. Returns normalised amplitudes summing
    to 1. This is the classic Singer-Seering shaper used on cranes.
    """
    zeta = crane.sway_damping
    w = crane.natural_frequency()
    wd = w * math.sqrt(max(1e-9, 1.0 - zeta * zeta))
    td = math.pi / wd  # half damped period
    k = math.exp(-zeta * math.pi / math.sqrt(max(1e-9, 1.0 - zeta * zeta)))
    a0 = 1.0 / (1.0 + k)
    a1 = k / (1.0 + k)
    return [(0.0, a0), (td, a1)]


# ── high-level traverse simulation ──────────────────────────────────────────


@dataclass
class TraverseResult:
    reached: bool
    settle_time_s: float
    residual_sway_m: float
    peak_sway_m: float
    final_x: float
    steps: int
    trajectory: List[List[float]] = field(default_factory=list)


def simulate_traverse(
    crane: GantryCrane,
    x_target: float,
    controller: AntiSwayController | None = None,
    dt: float = 1.0 / 50.0,
    max_time_s: float = 120.0,
    pos_tol_m: float = 0.10,
    sway_tol_m: float = 0.05,
    record: bool = False,
) -> TraverseResult:
    """Drive the trolley from rest at x=0 to ``x_target`` under anti-sway control.

    "Settled" = trolley within ``pos_tol_m`` of target AND the lateral load
    excursion (L·sinθ) within ``sway_tol_m`` AND sway rate near zero.
    Returns timing + residual-sway metrics used by the terminal's KPI gate.
    """
    if abs(x_target) > crane.rail_length:
        raise ValueError(f"x_target {x_target} exceeds rail_length {crane.rail_length}")
    ctrl = controller or AntiSwayController()
    state = [0.0, 0.0, 0.0, 0.0]
    L = crane.cable_length
    n = int(max_time_s / dt)
    peak = 0.0
    settle_time = -1.0
    traj: List[List[float]] = []
    for i in range(n):
        u = ctrl.command(crane, state, x_target)
        state = crane.step(state, u, dt)
        sway = abs(L * math.sin(state[2]))
        peak = max(peak, sway)
        if record:
            traj.append(list(state))
        settled = (
            abs(state[0] - x_target) <= pos_tol_m
            and sway <= sway_tol_m
            and abs(state[3]) <= 0.01
        )
        if settled and settle_time < 0.0:
            settle_time = (i + 1) * dt
            break
    residual = abs(L * math.sin(state[2]))
    return TraverseResult(
        reached=settle_time >= 0.0,
        settle_time_s=settle_time if settle_time >= 0.0 else max_time_s,
        residual_sway_m=residual,
        peak_sway_m=peak,
        final_x=state[0],
        steps=i + 1,
        trajectory=traj,
    )


def lift_cycle_time(
    crane: GantryCrane,
    traverse_m: float,
    hoist_up_m: float,
    hoist_down_m: float,
    hoist_speed_mps: float = 1.5,
) -> float:
    """Single-box cycle time estimate (s): hoist-up → traverse → hoist-down.

    Used by ``stow_plan`` to estimate berth productivity (moves/hour). The
    traverse term reuses the anti-sway settle time so the estimate reflects
    real sway-limited motion, not an unphysical bang-bang move.
    """
    res = simulate_traverse(crane, traverse_m)
    hoist = (hoist_up_m + hoist_down_m) / max(1e-6, hoist_speed_mps)
    return res.settle_time_s + hoist


def moves_per_hour(cycle_time_s: float) -> float:
    """Convert a per-box cycle time to the terminal productivity KPI."""
    if cycle_time_s <= 0:
        raise ValueError("cycle_time_s must be positive")
    return 3600.0 / cycle_time_s

File: 20-actors/niyaku/methods/crane_dynamics.clj
#!/usr/bin/env bb
;; Working Clojure port of methods/crane_dynamics.py.
(ns niyaku.methods.crane-dynamics
  "crane_dynamics — gantry / ship-to-shore (STS) crane anti-sway physics core (ADR-2606082000).

  The defining control problem of automated container handling is ANTI-SWAY: a quay crane moves
  a 20-40 t container suspended on cables while the trolley traverses 30-50 m ship→shore. The
  load is a pendulum; an aggressive move excites residual sway that must settle to < a few cm
  before the spreader can land the box. Cart + hanging payload — the same Cartpole topology
  isaac_sway_sim drives through the clean-room isaacsim.core.api surface.

  Analytic/control core: a physically-correct hanging pendulum (RK4), a state-feedback
  anti-sway controller, and a ZV input-shaper (the terminal-deployed technique). Pure compute,
  no NumPy; state is a plain [x x_dot theta theta_dot] vector. Equilibrium is the load hanging
  down (theta=0, STABLE — gravity restores).

  Run:  bb --classpath 20-actors 20-actors/niyaku/methods/crane_dynamics.clj"
  (:require [clojure.string :as str]))

(defn make-crane
  [& {:keys [cable-length gravity sway-damping accel-max velocity-max rail-length]
      :or {cable-length 30.0 gravity 9.81 sway-damping 0.02
           accel-max 0.6 velocity-max 4.0 rail-length 60.0}}]
  {:cable-length cable-length :gravity gravity :sway-damping sway-damping
   :accel-max accel-max :velocity-max velocity-max :rail-length rail-length})

(defn- clamp [v lim] (max (- lim) (min lim v)))

(defn natural-frequency
  "Undamped sway natural frequency ω = sqrt(g/L) (rad/s)."
  [c] (Math/sqrt (/ (:gravity c) (:cable-length c))))

(defn sway-period
  "Sway period T = 2π/ω (s) — sets the input-shaper impulse spacing."
  [c] (/ (* 2.0 Math/PI) (natural-frequency c)))

(defn derivatives
  "Continuous-time state derivative for [x x_dot theta theta_dot]. Full (non-linearised)
  hanging-pendulum-on-trolley with viscous sway damping; trolley accel = clamped command u."
  [c state u]
  (let [[_ x-dot theta theta-dot] state
        a (clamp u (:accel-max c))
        L (:cable-length c)
        g (:gravity c)
        zeta-w (* (:sway-damping c) (natural-frequency c))
        theta-acc (+ (* (- (/ g L)) (Math/sin theta))
                     (* (- (/ a L)) (Math/cos theta))
                     (* -2.0 zeta-w theta-dot))]
    [x-dot a theta-dot theta-acc]))

(defn- add4 [s k h] (mapv (fn [si ki] (+ si (* h ki))) s k))

(defn step
  "Advance one step by classic RK4 (stable for the stiff sway mode); enforce velocity envelope."
  [c state u dt]
  (let [k1 (derivatives c state u)
        k2 (derivatives c (add4 state k1 (/ dt 2.0)) u)
        k3 (derivatives c (add4 state k2 (/ dt 2.0)) u)
        k4 (derivatives c (add4 state k3 dt) u)
        nxt (mapv (fn [i] (+ (nth state i)
                             (* (/ dt 6.0)
                                (+ (nth k1 i) (* 2.0 (nth k2 i)) (* 2.0 (nth k3 i)) (nth k4 i)))))
                  (range 4))]
    (assoc nxt 1 (clamp (nth nxt 1) (:velocity-max c)))))

;; ── anti-sway state-feedback controller ─────────────────────────────────────
(defn make-controller
  [& {:keys [kp kd k-theta k-thetad] :or {kp 0.4 kd 1.7 k-theta 5.0 k-thetad 3.0}}]
  {:kp kp :kd kd :k-theta k-theta :k-thetad k-thetad})

(defn command
  "u = -kp·ω²(x-x_target) - kd·ω·x_dot + k_theta·θ + k_thetad/ω·θ_dot, clamped to accel-max.
  The sway terms (positive sign at θ=0 equilibrium) actively bleed pendulum energy."
  [ctrl c state x-target]
  (let [[x x-dot theta theta-dot] state
        w (natural-frequency c)
        u (+ (* (- (:kp ctrl)) w w (- x x-target))
             (* (- (:kd ctrl)) w x-dot)
             (* (:k-theta ctrl) theta)
             (* (/ (:k-thetad ctrl) w) theta-dot))]
    (clamp u (:accel-max c))))

;; ── ZV input shaper (open-loop anti-sway) ────────────────────────────────────
(defn zv-shaper
  "Zero-Vibration shaper impulses [[t0 a0] [t1 a1]] (Singer-Seering), amplitudes summing to 1."
  [c]
  (let [zeta (:sway-damping c)
        w (natural-frequency c)
        wd (* w (Math/sqrt (max 1e-9 (- 1.0 (* zeta zeta)))))
        td (/ Math/PI wd)
        k (Math/exp (/ (* (- zeta) Math/PI) (Math/sqrt (max 1e-9 (- 1.0 (* zeta zeta))))))
        a0 (/ 1.0 (+ 1.0 k))
        a1 (/ k (+ 1.0 k))]
    [[0.0 a0] [td a1]]))

;; ── high-level traverse simulation ──────────────────────────────────────────
(defn simulate-traverse
  "Drive the trolley from rest at x=0 to x-target under anti-sway control. 'Settled' = trolley
  within pos-tol-m AND lateral load excursion (L·sinθ) within sway-tol-m AND sway rate ~0."
  [c x-target & {:keys [controller dt max-time-s pos-tol-m sway-tol-m record]
                 :or {controller (make-controller) dt (/ 1.0 50.0) max-time-s 120.0
                      pos-tol-m 0.10 sway-tol-m 0.05 record false}}]
  (when (> (Math/abs (double x-target)) (:rail-length c))
    (throw (ex-info (str "x_target " x-target " exceeds rail_length " (:rail-length c))
                    {:type :value-error})))
  (let [L (:cable-length c)
        n (int (/ max-time-s dt))
        finalize (fn [state peak settle-time steps traj]
                   {:reached (>= settle-time 0.0)
                    :settle-time-s (if (>= settle-time 0.0) settle-time max-time-s)
                    :residual-sway-m (Math/abs (* L (Math/sin (nth state 2))))
                    :peak-sway-m peak
                    :final-x (nth state 0)
                    :steps steps
                    :trajectory (persistent! traj)})]
    (loop [i 0, state [0.0 0.0 0.0 0.0], peak 0.0, traj (transient [])]
      (if (>= i n)
        (finalize state peak -1.0 n traj)
        (let [u (command controller c state x-target)
              nstate (step c state u dt)
              sway (Math/abs (* L (Math/sin (nth nstate 2))))
              npeak (max peak sway)
              traj' (if record (conj! traj nstate) traj)
              settled (and (<= (Math/abs (- (nth nstate 0) x-target)) pos-tol-m)
                           (<= sway sway-tol-m)
                           (<= (Math/abs (nth nstate 3)) 0.01))]
          (if settled
            (finalize nstate npeak (* (+ i 1) dt) (+ i 1) traj')
            (recur (inc i) nstate npeak traj')))))))

(defn lift-cycle-time
  "Single-box cycle time (s): hoist-up → traverse (anti-sway settle time) → hoist-down."
  [c traverse-m hoist-up-m hoist-down-m & {:keys [hoist-speed-mps] :or {hoist-speed-mps 1.5}}]
  (let [res (simulate-traverse c traverse-m)
        hoist (/ (+ hoist-up-m hoist-down-m) (max 1e-6 hoist-speed-mps))]
    (+ (:settle-time-s res) hoist)))

(defn moves-per-hour
  "Convert a per-box cycle time to the terminal productivity KPI."
  [cycle-time-s]
  (when (<= cycle-time-s 0)
    (throw (ex-info "cycle_time_s must be positive" {:type :value-error})))
  (/ 3600.0 cycle-time-s))

(defn main [& _]
  (let [c (make-crane :cable-length 25.0 :accel-max 0.7)
        res (simulate-traverse c 30.0 :max-time-s 300.0)]
    (println (format "niyaku crane anti-sway: ω=%.4f rad/s  T=%.2fs" (natural-frequency c) (sway-period c)))
    (println (format "  traverse 30 m → reached=%s settle=%.2fs residual=%.4f m peak=%.4f m"
                     (:reached res) (double (:settle-time-s res))
                     (double (:residual-sway-m res)) (double (:peak-sway-m res))))
    (let [t (lift-cycle-time c 30.0 20.0 18.0)]
      (println (format "  cycle %.1fs → %.1f moves/hour" t (moves-per-hour t))))))

(when (= *file* (System/getProperty "babashka.file"))
  (main))
